### PR TITLE
Overhauling the Sound System

### PR DESCRIPTION
**Pull request by Jannify**
- Fixed sound volume calculation
- Added FMODEventInstance type which makes sounds stoppable (like the prawn drill)
- Moved sound whitelist to the launcher directory so it's accessible from client and server
- Changed FMOD classes naming schema
- Added function to suppress array of packet types

### CODE REVIEW

**Comment by tornac1234** [comment]
Welp what part of the code do you need me to redo ?

**Comment by Jannify** [comment]
Currently investigating a bug which result in sounds having short echos bc of packet loop.

**Comment by Jannify** [comment]
I couldn't get the echo anymore. Therefore this pr it is open for testing (best with two or more persons).

**Comment by tornac1234** [comment]
It's easier to check if you have many players connected and close to each other

File: NitroxPatcher/Patches/Dynamic/FMODUWE_PlayOneShotImpl_Patch.cs [hunk 1]
```diff
@@ -6,31 +6,30 @@
...
+    }
 
-        public override void Patch(Harmony harmony)
-        {
-            fmodSystem = NitroxServiceLocator.LocateService<FMODSystem>();
-            PatchMultiple(harmony, TARGET_METHOD, prefix:true, postfix:true);
-        }
+    public override void Patch(Harmony harmony)
+    {
+        fmodSystem = NitroxServiceLocator.LocateService<FMODSystem>();
```
Comment (tornac1234) [inline NitroxPatcher/Patches/Dynamic/FMODUWE_PlayOneShotImpl_Patch.cs:32]:
```suggestion
        fmodSystem = Resolve<FMODSystem>();
```

File: NitroxPatcher/Patches/Dynamic/FMOD_CustomEmitter_Start_Patch.cs [hunk 2]
```diff
@@ -6,62 +6,66 @@
...
+    }
 
-        public override void Patch(Harmony harmony)
-        {
-            fmodSystem = NitroxServiceLocator.LocateService<FMODSystem>();
-            PatchPrefix(harmony, TARGET_METHOD);
-        }
+    public override void Patch(Harmony harmony)
+    {
+        fmodSystem = NitroxServiceLocator.LocateService<FMODSystem>();
```
Comment (tornac1234) [inline NitroxPatcher/Patches/Dynamic/FMOD_CustomEmitter_Start_Patch.cs:68]:
```suggestion
        fmodSystem = Resolve<FMODSystem>();
```

File: NitroxPatcher/Patches/Dynamic/FMOD_CustomLoopingEmitter_PlayStopSound_Patch.cs [hunk 3]
```diff
@@ -5,39 +5,40 @@
...
         {
-            fmodSystem = NitroxServiceLocator.LocateService<FMODSystem>();
-            PatchMultiple(harmony, TARGET_METHOD, prefix:true, postfix:true);
+            fmodSystem.PlayCustomLoopingEmitter(nitroxEntity.Id, __instance.assetStop.path);
         }
     }
+
+    public override void Patch(Harmony harmony)
+    {
+        fmodSystem = NitroxServiceLocator.LocateService<FMODSystem>();
```
Comment (tornac1234) [inline NitroxPatcher/Patches/Dynamic/FMOD_CustomLoopingEmitter_PlayStopSound_Patch.cs:41]:
```suggestion
        fmodSystem = Resolve<FMODSystem>();
```

File: NitroxPatcher/Patches/Dynamic/FMOD_StudioEventEmitter_PlayUI_Patch.cs [hunk 4]
```diff
@@ -6,43 +6,44 @@
...
 
-        public override void Patch(Harmony harmony)
-        {
-            fmodSystem = NitroxServiceLocator.LocateService<FMODSystem>();
-            PatchMultiple(harmony, TARGET_METHOD, prefix:true, postfix:true);
-        }
-
+    public override void Patch(Harmony harmony)
+    {
+        fmodSystem = NitroxServiceLocator.LocateService<FMODSystem>();
```
Comment (tornac1234) [inline NitroxPatcher/Patches/Dynamic/FMOD_StudioEventEmitter_PlayUI_Patch.cs:45]:
```suggestion
        fmodSystem = Resolve<FMODSystem>();
```

File: NitroxPatcher/Patches/Dynamic/FMOD_StudioEventEmitter_Start_Patch.cs [hunk 5]
```diff
@@ -5,40 +5,41 @@
...
-            fmodSystem = NitroxServiceLocator.LocateService<FMODSystem>();
-            PatchPostfix(harmony, TARGET_METHOD);
+            fmodController = entity.gameObject.AddComponent<FMODEmitterController>();
         }
+        fmodController.AddEmitter(__instance.asset.path, __instance, radius);
+    }
+
+    public override void Patch(Harmony harmony)
+    {
+        fmodSystem = NitroxServiceLocator.LocateService<FMODSystem>();
```
Comment (tornac1234) [inline NitroxPatcher/Patches/Dynamic/FMOD_StudioEventEmitter_Start_Patch.cs:42]:
```suggestion
        fmodSystem = Resolve<FMODSystem>();
```

File: NitroxPatcher/Patches/Dynamic/FMOD_StudioEventEmitter_Stop_Patch.cs [hunk 6]
```diff
@@ -5,40 +5,41 @@
...
         {
-            fmodSystem = NitroxServiceLocator.LocateService<FMODSystem>();
-            PatchMultiple(harmony, TARGET_METHOD, prefix:true, postfix:true);
+            fmodSystem.PlayStudioEmitter(nitroxEntity.Id, __instance.asset.path, false, allowFadeout);
         }
+    }
 
+    public override void Patch(Harmony harmony)
+    {
+        fmodSystem = NitroxServiceLocator.LocateService<FMODSystem>();
```
Comment (tornac1234) [inline NitroxPatcher/Patches/Dynamic/FMOD_StudioEventEmitter_Stop_Patch.cs:41]:
```suggestion
        fmodSystem = Resolve<FMODSystem>();
```

File: NitroxPatcher/Patches/Dynamic/Utils_PlayFMODAsset_Patch.cs [hunk 7]
```diff
@@ -5,31 +5,30 @@
...
+    }
 
-        public override void Patch(Harmony harmony)
-        {
-            fmodSystem = NitroxServiceLocator.LocateService<FMODSystem>();
-            PatchMultiple(harmony, TARGET_METHOD, prefix:true, postfix:true);
-        }
+    public override void Patch(Harmony harmony)
+    {
+        fmodSystem = NitroxServiceLocator.LocateService<FMODSystem>();
```
Comment (tornac1234) [inline NitroxPatcher/Patches/Dynamic/Utils_PlayFMODAsset_Patch.cs:31]:
```suggestion
        fmodSystem = Resolve<FMODSystem>();
```

**Review by tornac1234 — COMMENTED** [review]
Just found out those, not sure if that's an oversight or voluntary

File: NitroxClient/GameLogic/RemotePlayer.cs [hunk 8]
```diff
@@ -284,7 +290,7 @@ private void UpdateEquipmentVisibility()
         }
         
         /// <summary>
-        /// Allows the remote player model to have its lightings dynamicly adjusted
+        /// Allows the remote player model to have its lightnings dynamically adjusted
```
Comment (tornac1234) [inline NitroxClient/GameLogic/RemotePlayer.cs:293]: Not sure why you talk about lightnings, since the SkyApplier is meant for lighting

**Comment by Jannify** [comment]
Is based on #1908 so should be merged after that

**Review by Measurity — CHANGES_REQUESTED** [review]

File: NitroxModel/Helper/NitroxUser.cs [hunk 9]
```diff
@@ -14,21 +14,12 @@ public static class NitroxUser
...
-            () => Environment.GetEnvironmentVariable(LAUNCHER_PATH_ENV_KEY),
-            () =>
-            {
-                Assembly currentAsm = Assembly.GetEntryAssembly();
-                if (currentAsm?.GetName().Name.Equals("NitroxLauncher") ?? false)
-                {
-                    return Path.GetDirectoryName(currentAsm.Location);
-                }
-                return null;
-            }
```
Comment (Measurity) [inline NitroxModel/Helper/NitroxUser.cs:31]: Removing this makes the game fail to find the launcher path (and not load Nitrox) for me.

**Review by Measurity — CHANGES_REQUESTED** [review]

File: NitroxModel/Helper/NitroxUser.cs (hunk 9 above)
Comment (Jannify) [reply NitroxModel/Helper/NitroxUser.cs:31]: mb, this is a rebase artifact I missed

File: NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+               .Insert(
+                   new CodeInstruction(OpCodes.Ldarg_0),
+                   new CodeInstruction(OpCodes.Ldloc_0),
+                   new CodeInstruction(OpCodes.Ldarg_1),
+                   new CodeInstruction(OpCodes.Call, Reflect.Method(() => CalculateVolume(default, default, default, default)))
+               )
+               .InstructionEnumeration();
+    }
+
+    // This method is called very often and should therefor be performant
```
Comment (tornac1234) [inline NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs:44]:
```suggestion
    // This method is called very often and should therefore be performant
```

File: NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    private static float CalculateVolume(float originalVolume, FootstepSounds instance, FMODAsset asset, Transform xform)
+    {
+        // For exosuits only "event:/sub/exo/step" is used inside FootstepSounds
+        if (asset.path.Equals("event:/sub/exo/step", StringComparison.Ordinal) && (!Player.main.currentMountedVehicle || Player.main.currentMountedVehicle.gameObject != instance.gameObject))
```
Comment (tornac1234) [inline NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs:48]: Path string could be moved on top of the file to also be used by ``stepSoundRadius``

File: NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+using FMOD.Studio;
+using HarmonyLib;
+using NitroxClient.GameLogic.FMOD;
+using NitroxModel.GameLogic.FMOD;
+using NitroxModel.Helper;
+using UnityEngine;
+
+namespace NitroxPatcher.Patches.Dynamic;
+
+public sealed partial class FootstepSounds_OnStep_Patch : NitroxPatch, IDynamicPatch
```
Comment (tornac1234) [inline NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs:14]: Should add a test for the transpiler

File: NitroxPatcher/Patches/Dynamic/Utils_PlayFMODAsset_Patch.cs [hunk 13]
```diff
@@ -1,7 +1,8 @@
-﻿using System.Reflection;
+using System.Reflection;
 using NitroxClient.GameLogic.FMOD;
 using NitroxModel.Helper;
 using NitroxModel_Subnautica.DataStructures;
+using NitroxModel.GameLogic.FMOD;
```
Comment (tornac1234) [inline NitroxPatcher/Patches/Dynamic/Utils_PlayFMODAsset_Patch.cs:5]: Sort usings

File: NitroxServer/Communication/Packets/Processors/FMODAssetProcessor.cs [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        this.playerManager = playerManager;
+        this.fmodWhitelist = fmodWhitelist;
+    }
+
+    public override void Process(FMODAssetPacket packet, Player sendingPlayer)
+    {
+        if (!fmodWhitelist.TryGetSoundData(packet.AssetPath, out SoundData soundData))
+        {
+            Log.Error($"[FMODAssetProcessor] Whitelist has no item for {packet.AssetPath}.");
```
Comment (tornac1234) [inline NitroxServer/Communication/Packets/Processors/FMODAssetProcessor.cs:24]: nameof

File: NitroxServer/Communication/Packets/Processors/FMODEventInstanceProcessor.cs [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        this.playerManager = playerManager;
+        this.fmodWhitelist = fmodWhitelist;
+    }
+
+    public override void Process(FMODEventInstancePacket packet, Player sendingPlayer)
+    {
+        if (!fmodWhitelist.TryGetSoundData(packet.AssetPath, out SoundData soundData))
+        {
+            Log.Error($"[FMODAssetProcessor] Whitelist has no item for {packet.AssetPath}.");
```
Comment (tornac1234) [inline NitroxServer/Communication/Packets/Processors/FMODEventInstanceProcessor.cs:24]:
```suggestion
            Log.Error($"[{nameof(FMODEventInstanceProcessor)}] Whitelist has no item for {packet.AssetPath}.");
```

File: NitroxServer/Communication/Packets/Processors/FMODEventInstanceProcessor.cs [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+            return;
+        }
+
+        foreach (Player player in playerManager.GetConnectedPlayers())
+        {
+            float distance = NitroxVector3.Distance(player.Position, packet.Position);
+            if (player != sendingPlayer &&
+                (soundData.IsGlobal || player.SubRootId.Equals(sendingPlayer.SubRootId)) &&
+                distance <= soundData.Radius)
```
Comment (tornac1234) [inline NitroxServer/Communication/Packets/Processors/FMODEventInstanceProcessor.cs:33]:
```suggestion
                distance < soundData.Radius)
```
According to SoundHelper.CalculateVolume

**Review by tornac1234 — CHANGES_REQUESTED** [review]
Haven't tested in game yet, also didn't understand how FMODCustomLoopingEmitter works as we only broadcast its start but not its stop event ?

**Comment by tornac1234** [comment]
In-game tests will be required

**Comment by dartasen** [comment]
Can organize a test session once this PR is rebased

**Review by dartasen — APPROVED** [review]

File: NitroxPatcher/Patches/Dynamic/SeaMoth_onLightsToggled_Patch.cs [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+
+public sealed partial class SeaMoth_onLightsToggled_Patch : NitroxPatch, IDynamicPatch
+{
+    private static readonly MethodInfo TARGET_METHOD = Reflect.Method((SeaMoth t) => t.onLightsToggled(default(bool)));
+
+    public static void Postfix(SeaMoth __instance)
+    {
+        if (__instance.TryGetIdOrWarn(out NitroxId id))
+        {
+            SeamothMetadata metadata = Resolve<SeamothMetadataExtractor>().Extract(__instance);
```
Comment (tornac1234) [inline NitroxPatcher/Patches/Dynamic/SeaMoth_onLightsToggled_Patch.cs:18]: This is an unconventional way of getting the metadata but still works

**Review by tornac1234 — COMMENTED** [review]
Only minor stuff

**Review by tornac1234 — APPROVED** [review]

**Review by dartasen — APPROVED** [review]

**Pull request merged**